### PR TITLE
APM-1492 Initial Django Implementation

### PR DESCRIPTION
**Pull request by tylerwiersing**

### CODE REVIEW

File: auklet/broker.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+        topic_suffix = "{}/{}".format(
+            self.client.org_id, self.client.app_id)
+        self.producer_types = {
+            "monitoring": "python/profiler/{}".format(topic_suffix),
+            "event": "python/events/{}".format(topic_suffix),
+        }
+
+    def _write_conf(self, info):
+        with open(self.com_config_filename, "w") as conf:
+            conf.write(json.dumps(info))
```
Comment (bleib1dj) [inline auklet/broker.py]: What’s in this conf?

File: auklet/client.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+import msgpack
+
+from time import time
+from uuid import uuid4
+
+from django.conf import settings
+
+from .errors import AukletConfigurationError
+from .broker import MQTTClient
+from .stats import Event, SystemMetrics, FilenameCaches
```
Comment (bleib1dj) [inline auklet/client.py]: Should we be using absolute path rather than relative like we do in the django app now?

File: auklet/client.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    Get an Auklet Client
+    """
+    global _client
+    if _client is None:
+        _client = DjangoClient()
+    return _client
+
+
+class DjangoClient(object):
+    identification_filename = ".auklet/identification"
```
Comment (bleib1dj) [inline auklet/client.py]: What’s in this identification file?

File: auklet/monitoring/__init__.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+from auklet.monitoring.logging import AukletLogging
+from auklet.monitoring.processing import Client
+from auklet.stats import MonitoringTree
+
+__all__ = ['Monitoring']
+
+except_hook_set = False
+
+
+class Monitoring(AukletLogging):
```
Comment (bleib1dj) [inline auklet/monitoring/__init__.py]: If this is the class people initiate I’d like to see them start doing `Auklet(...)` rather than `Monitoring(...)`

File: auklet/monitoring/__init__.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    timer = signal.ITIMER_REAL
+    sig = signal.SIGALRM
+    stopping = False
+    stopped = False
+
+    interval = 1e-3  # 1ms
+
+    total_samples = 0
+
+    emission_rate = 60000  # 60 seconds
```
Comment (bleib1dj) [inline auklet/monitoring/__init__.py]: Shouldn’t emission rate be removed and things just be transmitted after a request completes?

File: auklet/monitoring/processing.py [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+S_TO_MS = 1000
+
+
+class Client(object):
+    producer_types = None
+    brokers = None
+    commit_hash = None
+    mac_hash = None
+    offline_filename = ".auklet/local.txt"
+    limits_filename = ".auklet/limits"
```
Comment (bleib1dj) [inline auklet/monitoring/processing.py]: Don’t need limits, offline management, or usage for Django and server apps

File: auklet/monitoring/__init__.py (hunk 5 above)
Comment (bleib1dj) [reply auklet/monitoring/__init__.py]: Also maybe this isn’t used now cause we’re only doing events for mvp? If so should this be removed?

File: auklet/monitoring/__init__.py (hunk 4 above)
Comment (tylerwiersing) [reply auklet/monitoring/__init__.py]: They don't do anything with this, in fact it will be removed

File: auklet/monitoring/__init__.py (hunk 5 above)
Comment (tylerwiersing) [reply auklet/monitoring/__init__.py]: Will be removed

File: auklet/monitoring/processing.py (hunk 6 above)
Comment (tylerwiersing) [reply auklet/monitoring/processing.py]: Will be removed

File: auklet/client.py (hunk 2 above)
Comment (tylerwiersing) [reply auklet/client.py]: Yes these are leftover from some testing I was doing

File: auklet/broker.py (hunk 1 above)
Comment (tylerwiersing) [reply auklet/broker.py]: It's where we store the MQTT broker and port info

File: auklet/client.py (hunk 3 above)
Comment (tylerwiersing) [reply auklet/client.py]: This is where normally we'd store stuff like the `client_id` and `client_password` for the other agents, but here it only holds the `org_id`

File: auklet/licenses/profiling [hunk 7]
```diff
@@ -0,0 +1,27 @@
+Copyright (c) 2014-2015, What! Studio
```
Comment (donaldwasserman) [inline auklet/licenses/profiling]: Is this right?

File: auklet/broker.py (hunk 1 above)
Comment (bleib1dj) [reply auklet/broker.py]: Do they change at all/could the info be defaulted for prod envy and be modified with env vars. Just concerned about file manipulation by multiple threads at once. Especially writes.

File: auklet/client.py (hunk 3 above)
Comment (bleib1dj) [reply auklet/client.py]: Kk and what is org id used for again?

File: auklet/licenses/profiling (hunk 7 above)
Comment (tylerwiersing) [reply auklet/licenses/profiling]: We had to include this bc some of the code we were using was from another package I believe, @rjenkinsjr could confirm

File: auklet/client.py (hunk 3 above)
Comment (tylerwiersing) [reply auklet/client.py]: It is part of the MQTT topic that we produce to along with the broker username(app_id)

File: auklet/broker.py (hunk 1 above)
Comment (tylerwiersing) [reply auklet/broker.py]: Don't think we can really modify this with env vars, the write only happens once and it is only when the django app first starts up, after all apps are ready.

Comment (bleib1dj) [reply auklet/broker.py]: @tylerwiersing what is the example json that’s stored?

File: auklet/client.py (hunk 3 above)
Comment (bleib1dj) [reply auklet/client.py]: Kk rather just have them provide that as a required Django setting var

File: auklet/licenses/profiling (hunk 7 above)
Comment (bleib1dj) [reply auklet/licenses/profiling]: @tylerwiersing is that code still present after the refactor?

File: auklet/broker.py (hunk 1 above)
Comment (tylerwiersing) [reply auklet/broker.py]: `{"brokers": "mqtt.broker.io", "port": "8888"}`

File: auklet/licenses/profiling (hunk 7 above)
Comment (tylerwiersing) [reply auklet/licenses/profiling]: No, we can remove this

File: auklet/broker.py (hunk 1 above)
Comment (bleib1dj) [reply auklet/broker.py]: Yeah lets just set those as the defaults and make it so they can be overwritten by django setting vars.

Comment (bleib1dj) [reply auklet/broker.py]: Also looks like we’re pulling a ca cert still?

Comment (bleib1dj) [reply auklet/broker.py]: @tylerwiersing ^

File: README.rst [hunk 8]
```diff
@@ -0,0 +1,140 @@
+.. raw:: html
+    <p align="center">
+
+.. image:: https://s3.amazonaws.com/auklet/static/auklet_python.png
+    :target: https://auklet.io
+    :align: center
+    :width: 1000
+    :alt: Auklet - Problem Solving Software for Python
```
Comment (bleib1dj) [inline README.rst]: Need a new image from Shawn with the django label

File: README.rst [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+
+
+Release Tracking
+^^^^^^^^^^^^^^^^
+To track releases and identify which devices are running what version of code
+we currently require that you provide the commit hash of your deployed code.
+This value needs to be passed into the constructor through `release`.
+The value needs to be the commit hash that represents the
+deployed version of your application. There are a couple ways for which to set
+this based upon the style of deployment of your application.
```
Comment (bleib1dj) [inline README.rst]: This could be where we indicate adding release and version to the config dict. But the paragraph needs to be updated to not say they are reqs.

File: README.rst [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+* `Auklet`_
+* `Python Documentation`_
+* `Issue Tracker`_
+
+.. _Auklet: https://auklet.io
+.. _hello@auklet.io: mailto:hello@auklet.io
+.. _ESG-USA: https://github.com/ESG-USA
+.. _ESG Organization: https://github.com/ESG-USA
+.. _Python Documentation: https://docs.auklet.io/docs/python-integration
+.. _Issue Tracker: https://github.com/aukletio/Auklet-Agent-Python/issues
```
Comment (bleib1dj) [inline README.rst]: Please review all these

File: auklet/broker.py [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+        try:
+            try:
+                res = urlopen(url)
+            except HTTPError as e:
+                # Allow for accessing redirect w/o including the
+                # Authorization token.
+                res = urlopen(e.geturl())
+        except URLError:
+            return False
+        filename = ".auklet/ca.pem"
```
Comment (bleib1dj) [inline auklet/broker.py]: Probs wanna make sure this dir exists. Also please confirm we indicate the django app must have write access to the directory it is executing in.

File: auklet/utils.py [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+    return hashlib.md5(b(mac)).hexdigest()
+
+
+def get_commit_hash():
+    try:
+        with open(".auklet/version", "r") as auklet_file:
+            return auklet_file.read().rstrip()
+    except IOError:
+        # TODO Error out app if no commit hash
+        return ""
```
Comment (bleib1dj) [inline auklet/utils.py]: This fxn is outdated for django

File: auklet/utils.py [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+
+
+def get_abs_path(path):
+    try:
+        return os.path.abspath(path).split('/.auklet')[0]
+    except IndexError:
+        return ''
```
Comment (bleib1dj) [inline auklet/utils.py]: Still need?

File: README.rst [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+
+.. image:: https://api.codeclimate.com/v1/badges/7c2cd3bc63a70ac7fd73/maintainability
+   :target: https://codeclimate.com/repos/5a54e10be3d6cb4d7d0007a8/maintainability
+   :alt: Code Climate Maintainability
+
+.. image:: https://api.codeclimate.com/v1/badges/7c2cd3bc63a70ac7fd73/test_coverage
+   :target: https://codeclimate.com/repos/5a54e10be3d6cb4d7d0007a8/test_coverage
+   :alt: Test Coverage
+
+
```
Comment (bleib1dj) [inline README.rst]: What versions of Django do we support?

File: README.rst [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+This is the official Django agent for `Auklet`_, official supports 2.7.9+ and 3.4-3.7, and
```
Comment (bleib1dj) [inline README.rst]: Think we can drop the python version info here since it’s in the tag above. Need to instead call out django versions we work with. If we need to setup a similar tox setup we can do that in a different PR and @joshuahufford can assist

File: auklet/broker.py (hunk 1 above)
Comment (tylerwiersing) [reply auklet/broker.py]: Correct we are pulling the ca cert

File: README.rst [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+
+.. sourcecode:: python
+
+    import os
+    from django.core.wsgi import get_wsgi_application
+    from django.conf import settings
+    from auklet.middleware import WSGIAukletMiddleware
+
+    application = get_wsgi_application()
+    application = WSGIAukletMiddleware(application)
```
[the remark this answers is not in the document]
Comment (bleib1dj) [reply README.rst]: If we were to call out that the Auklet middleware must be listed first in the list of middleware achieve the same outcome?

[the remark this answers is not in the document]
Comment (tylerwiersing) [reply README.rst]: I don't think we really need to provide any additional context around this, it is an option for people to use if-for their specific use case-it is easier/better to use a wsgi middleware than builtin django ones. They should not include both of these setups though, either default django middleware list or wsgi or else they will produce 2x event packets. It's basically a convenience option for some, I would imagine most people would use the builtin django middleware but the option of wsgi middleware is also provided by opbeat and sentry

[the remark this answers is not in the document]
Comment (tylerwiersing) [reply README.rst]: It would probably be good to add in "don't use both"

File: README.rst [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+)
+
+To set up default django middleware error handling add the auklet middleware
+to the end of your middleware configs:
+
+.. sourcecode:: python
+
+    MIDDLEWARE = (
+        ...,
+        "auklet.middleware.AukletMiddleware",
```
Comment (bleib1dj) [inline README.rst]: Should this normally go before the other middleware rather than after them?

Comment (tylerwiersing) [reply README.rst]: It doesn't really matter in the end, there are some unique circumstances that I can add to the readme about ordering and other custom middleware they use

File: README.rst [hunk 18]
```diff
@@ -0,0 +1,141 @@
...
+Resources
+---------
+* `Auklet`_
+* `Python Documentation`_
+* `Issue Tracker`_
+
+.. _Auklet: https://auklet.io
+.. _hello@auklet.io: mailto:hello@auklet.io
+.. _ESG-USA: https://github.com/ESG-USA
+.. _ESG Organization: https://github.com/ESG-USA
```
Comment (bleib1dj) [inline README.rst]: Can remove the ESG references

File: README.rst [hunk 19]
```diff
@@ -0,0 +1,141 @@
...
+Authorization
+^^^^^^^^^^^^^
+To authorize your application you need to provide both an API key and app id.
+These values are available in the connection settings of your application as
+well as during initial setup.
+
+
+Release Tracking
+^^^^^^^^^^^^^^^^
+To track releases and identify which devices are running what version of code
```
Comment (bleib1dj) [inline README.rst]: Let’s start this off with “Optionally, you can track releases and identify which servers are running what variant of code. To do this you may provide...” just to make it more apparent this section is optional.

File: auklet/broker.py [hunk 20]
```diff
@@ -0,0 +1,88 @@
...
+            self.client.org_id, self.client.app_id)
+        self.producer_types = {
+            "monitoring": "python/profiler/{}".format(topic_suffix),
+            "event": "python/events/{}".format(topic_suffix),
+        }
+
+    def _get_certs(self):
+        url = Request(
+            build_url(self.client.base_url, "private/devices/certificates/"),
+            headers={"Authorization": "JWT %s" % self.client.apikey})
```
Comment (bleib1dj) [inline auklet/broker.py]: Let’s use the new style of string formatting to be consistent.

File: auklet/utils.py [hunk 21]
```diff
@@ -0,0 +1,88 @@
...
+from auklet.__about__ import __version__ as auklet_version
+
+__all__ = ['open_auklet_url', 'create_file', 'clear_file', 'build_url',
+           'get_mac', 'get_device_ip', 'get_agent_version',
+           'post_auklet_url', 'create_dir', 'b', 'u']
+
+
+def open_auklet_url(url, apikey):
+    try:
+        res = requests.get(url, headers={"Authorization": "JWT %s" % apikey})
```
Comment (bleib1dj) [inline auklet/utils.py]: String formatting

File: auklet/utils.py [hunk 22]
```diff
@@ -0,0 +1,88 @@
...
+        return None
+    return res
+
+
+def post_auklet_url(url, apikey, data):
+    try:
+        res = requests.post(
+            url,
+            json=data,
+            headers={"Authorization": "JWT %s" % apikey,
```
Comment (bleib1dj) [inline auklet/utils.py]: String formatting

File: auklet/broker.py [hunk 23]
```diff
@@ -0,0 +1,88 @@
...
+    password = None
+    com_config_filename = ".auklet/communication"
+    port = 8883
+    producer_types = {
+        "monitoring": "python/profiler/{}/{}",
+        "event": "python/events/{}/{}",
+    }
+
+    def __init__(self, client):
+        self.client = client
```
Comment (bleib1dj) [inline auklet/broker.py]: Might be missing something but I don’t see self.client used anywhere. Is there any reason we need to pass in the entire object as opposed to just the broker and port?

Comment (tylerwiersing) [reply auklet/broker.py]: It also gets the `org_id` `app_id` and `apikey` from the client

Comment (bleib1dj) [reply auklet/broker.py]: I would still recommend passing those by themselves to avoid tight coupling with the other class

File: README.rst [hunk 24]
```diff
@@ -0,0 +1,139 @@
...
+
+.. sourcecode:: python
+
+    MIDDLEWARE = (
+        ...,
+        "auklet.middleware.AukletMiddleware",
+    )
+
+NOTE: If you are already using an error handling middleware which returns a response you need to disable it or add:
+.. sourcecode:: python
```
Comment (shogun656) [inline README.rst]: I think you need to add one more line of whitespace before this line as it's not being formatted correctly

File: README.rst [hunk 25]
```diff
@@ -0,0 +1,139 @@
...
+To setup Auklet monitoring for you application simply include it in your
+`INSTALLED_APPS`:
+
+.. sourcecode:: python
+
+    INSTALLED_APPS = (
+    ...,
+    'auklet',
+    ...,
+)
```
Comment (shogun656) [inline README.rst]: This should be tabbed as it's not being formatted correctly

File: advanced_setup.rst [hunk 26]
```diff
@@ -0,0 +1,12 @@
+If you wish set up Auklet using a wsgi middleware instead of the
```
Comment (shogun656) [inline advanced_setup.rst]: if you wish `to` set up Auklet
